### PR TITLE
Ports - Jillianne & Shamira

### PR DESCRIPTION
**Pull request by jillirami**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? |  It was a bit challenging getting up to speed with the existing codebase. Next time we'll take more time to understand the individual responsibilities of each file before diving into the project instead of learning as we go.
What _inheritance_ relations exist between classes? | CsvRecord was the super class for the Trip class, Passenger class, and Driver class.
What _composition_ relations exist between classes? | Passenger has CsvRecord, Driver has CsvRecord, and Trip has CsvRecord. TripDispatch has passenger, trip and driver. 
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | Creating a helper method was a major decision for us, and we discussed making this a private method, but worked through the logic of calling our method as a helper and instead called it within a preexisting method for desired output.
Give an example of a _template method_ that you implemented for this assignment | The CsvRecord has a self.load all which utilizes the template method. 
Give an example of a _nominal test_ that you wrote for this assignment | It "created a new instance of a trip" is a nominal test we wrote.
Give an example of an _edge case test_ that you wrote for this assignment | It "does not provide negative revenue" if a ride is less than $1.65.
What is a concept that you gained more clarity on as you worked on this assignment | Inheritance of subclasses from super classes.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -3,7 +3,7 @@ source 'http://rubygems.org'
 ruby '2.5.1'
 
 gem 'rake'
-
+gem 'pry'
```
Comment (CheezItMan) [inline Gemfile:6]: When you're ready to submit, it's a good idea to remove pry and all the `binding.pry`.  It just makes your code a bit more tidy.  Not a big issue.

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
+require_relative "csv_record"
+require "csv"
+
+module RideShare
+  class Driver < CsvRecord
+    attr_reader :id, :name, :vin, :trips
+    attr_accessor :status
+
+    STATUS = [:AVAILABLE, :UNAVAILABLE]
```
Comment (CheezItMan) [inline lib/driver.rb:9]: This name doesn't sound quite right.  A more descriptive name would be:  `VALID_STATUS_VALUES`

File: lib/test.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+require_relative "passenger"
```
Comment (CheezItMan) [inline lib/test.rb:1]: It looks like this file shouldn't belong here.

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -131,10 +132,97 @@
...
+      )
+      @driver.add_trip(trip2)
+    end
+
+    it "returns a float" do
+      expect(@driver.total_revenue).must_be_kind_of Float
+    end
+
+    it "does not provide negative revenue" do
+      expect(@driver.total_revenue).must_equal 5.08
```
Comment (CheezItMan) [inline specs/driver_spec.rb:169]: You should have a test where the cost would make the total revenue go negative, like where the cost is less than 1.65.

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -131,10 +132,97 @@
...
+        passenger_id: 3,
+        start_time: "2016-08-08",
+        end_time: "2016-08-09",
+        rating: 1,
+      )
+      @driver.add_trip(trip)
+      ending_length = @driver.trips.length
+      puts "ending length", ending_length
+
+      expect(ending_length).must_equal starting_length + 1
```
Comment (CheezItMan) [inline specs/driver_spec.rb:225]: You should also check to see if the new trip is in the `@driver.trips` array?

File: specs/passenger_spec.rb [hunk 6]
```diff
@@ -67,4 +73,45 @@
...
+      @passenger.add_trip(trip1)
+      @passenger.add_trip(trip2)
+    end
+
+    it "each passengers total trip costs" do
+      expect(@passenger.net_expenditures).must_equal 25
+    end
+
+    it "total amount of time each passenger has spent on trips" do
+      expect(@passenger.total_time_spent).must_equal 172800
```
Comment (CheezItMan) [inline specs/passenger_spec.rb:114]: You should also have tests for a passenger without any trips.

You also have no tests for passengers with in-progress trips.

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -122,4 +121,47 @@ def build_test_dispatcher
...
+      test_passenger = @dispatcher.passengers[3]
+      test_passenger_id = test_passenger.id
+      @dispatcher.request_trip(test_passenger_id)
+      ending_length = @dispatcher.trips.length
+      expect(ending_length).must_equal starting_length + 1
+    end
+
+    it "sets the driver's status to :UNAVAILABLE and adds to driver class" do
+      trip = @dispatcher.request_trip(5)
+      expect(trip.driver.status).must_equal :UNAVAILABLE
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:140]: You should also test that the driver *was* available before the trip was requested.

File: lib/driver.rb [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        raise ArgumentError, "Invalid status #{@status}."
+      end
+    end
+
+    def add_trip(trip)
+      @trips << trip
+    end
+
+    def average_rating
+      rating = 0
```
Comment (CheezItMan) [inline lib/driver.rb:33]: What about handling in-progress trips?

File: lib/driver.rb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      @trips.each do |trip|
+        rating += trip.rating.to_f
+      end
+
+      return @trips.length > 0 ?
+               (rating / @trips.length) : rating
+    end
+
+    def total_revenue
+      revenue = 0
```
Comment (CheezItMan) [inline lib/driver.rb:44]: Similar to the above, no handling of in-progress trips.

File: lib/passenger.rb [hunk 10]
```diff
@@ -1,29 +1,44 @@
...
       @phone_number = phone_number
-      @trips = trips || []
+      @trips = trips
     end
 
     def add_trip(trip)
       @trips << trip
     end
 
+    def net_expenditures
```
Comment (CheezItMan) [inline lib/passenger.rb:20]: What about in-progress trips?

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly |  Don't use waves in your commit messages and instead just describe the functionality you added to your code.  You do have a good number of commits.
Answer comprehension questions | In the composition section you are confusing _inheritance_ with _composition_.  A `Passenger` instance _is-a_ `CsvRecord`.  That's a relationship by inheritance.  Instead a `Passenger` _has-a_ `Trip` would be an example of composition.  The `Passenger` has a `Trip` as an instance variable.  Also helper methods are generally private because users of your class do not call it.  The helper method is only called by other methods in the class.  Good example of an edge-case test.  
**Wave 1** |
Appropriate use of Ruby's `Time` | Check
`Trip` has a helper method to calculate duration | Check
`Passenger` methods to calculate total cost and total time | Check
Complete tests for new behavior | Yes, although you should test the edge-case of having no trips.
**Wave 2** |
`Driver` inherits from `CsvRecord`, and implements `from_csv` | Check
`Driver`s and `Trip`s are connected by `TripDispatcher` | Check
`Driver` has method to calculate average rating | Check
`Driver` has a method to calculate total revenue | Check
**Wave 3** |
`TripDispatcher` has a new method to create trips | Check
Trip creation logic is correctly implemented | Check, well done
New `Trip` is connected to the appropriate `Driver` and `Passenger` | Check
Tests for `request_trip` | You have some good tests, but I made some remarks inline in your code.
Wave 1 and 2 methods ignore in-progress trips | Check
Tests for Wave 1 and 2 methods w/ in-progress trips |  **Missing** no tests for in-progress trips.
**Overall** | You hit the main learning goals here, but see my comments, especially about the tests that were missing.  Overall, well done.
